checker/src/org/checkerframework/checker/index/upperbound/UpperBoundVisitor.java

```diff
@@ -33,6 +52,150 @@ public UpperBoundVisitor(BaseTypeChecker checker) {
+     *   <li>accesses of a public final field whose access expression (sometimes called the
+     *       receiver) is composed only of expressions in this list
+     * </ul>
+     *
+     * <p>Any other expression results in a warning.
+     */
+    @Override
+    public Void visitVariable(VariableTree node, Void p) {
+        AnnotatedTypeMirror atm = atypeFactory.getAnnotatedTypeLhs(node);
+        AnnotationMirror anm = atm.getAnnotationInHierarchy(atypeFactory.UNKNOWN);
+        if (anm != null && AnnotationUtils.hasElementValue(anm, "value")) {
+            // This is a dependent annotation. If this is a local variable,
+            // issue a warning; dependent annotations should not be written on local variables.
+            Element elt = TreeUtils.elementFromDeclaration(node);
+            boolean allVariablesFinal = true;
+
+            UBQualifier qual = UBQualifier.createUBQualifier(anm);
+            if (qual.isLessThanLengthQualifier()) {
+                LessThanLengthOf ltl = (LessThanLengthOf) qual;
+                for (String array : ltl.getSequences()) {
+                    Receiver rec = getReceiverForCheckingDependentTypes(array, node);
+                    // covers final and effectively final variables
+                    boolean isFinal =
+                            rec == null
+                                    || rec.isUnmodifiableByOtherCode()
+                                    || rec instanceof FlowExpressions.LocalVariable;
+                    if (!isFinal) {
+                        allVariablesFinal = false;
+                        checkIfPermittedInDependentTypeAnno(rec, array, node);
+                    }
+                }
+            }
+
+            if (elt.getKind() == ElementKind.LOCAL_VARIABLE && !allVariablesFinal) {
+                checker.report(Result.warning(LOCAL_VAR_ANNO), node);
+            }
+        }
+
+        return super.visitVariable(node, p);
+    }
+
+    private Receiver getReceiverForCheckingDependentTypes(String expr, Tree tree) {
+        try {
+            return atypeFactory.getReceiverFromJavaExpressionString(expr, getCurrentPath());
+        } catch (FlowExpressionParseUtil.FlowExpressionParseException e) {
+            // issue warning
+            checker.report(
+                    Result.warning(DEPENDENT_NOT_PERMITTED, expr, "the expression was unparseable"),
+                    tree);
+            return null;
+        }
+    }
+
+    private boolean checkIfPermittedInDependentTypeAnno(Receiver rec, String expr, Tree tree) {
+
+        if (rec == null) {
+            return false;
+        }
+
+        if (rec.isUnmodifiableByOtherCode() || rec instanceof FlowExpressions.LocalVariable) {
+            return true;
+        }
+
+        if (rec instanceof FlowExpressions.FieldAccess) {
+            FlowExpressions.FieldAccess faRec = (FlowExpressions.FieldAccess) rec;
+            if (faRec.getField().getModifiers().contains(Modifier.PRIVATE)
```

Comment:
Why isn't the receiver of a private field checked to see if it is permitted?

Reply:
Because private fields are always permitted in dependent annotations. In order to write a private field in a dependent annotation, we need to be in the class that the field is defined in, so all the code that might side-effect the field will be visited.